// ngx-build-plus/schematics/webcomponents-polyfill/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const workspace_1 = require("../utils/workspace");
const tasks_1 = require("@angular-devkit/schematics/tasks");
const spawn = require('cross-spawn');
// function npmInstall(options: any): Rule {
//   return function (tree: Tree, context: SchematicContext) {
//     spawn.sync('npm', ['install', options.package, options.switch], { stdio: 'inherit' });
//     return tree;
//   }
// }
function npmRun(options) {
    return function (tree, context) {
        spawn.sync('npm', ['run', options.script], { stdio: 'inherit' });
        return tree;
    };
}
exports.npmRun = npmRun;
function addWebComponentsPolyfill(_options) {
    return (tree, _context) => {
        const packageJson = loadPackageJson(tree);
        if (!packageJson['dependencies'] || !packageJson['dependencies']['@webcomponents/webcomponentsjs']) {
            _context.addTask(new tasks_1.NodePackageInstallTask({
                packageName: '@webcomponents/webcomponentsjs',
            }));
        }
        return schematics_1.chain([
            addScriptsToProject(tree, _options)
        ]);
    };
}
exports.addWebComponentsPolyfill = addWebComponentsPolyfill;
function addScriptsToProject(tree, options) {
    const workspace = workspace_1.getWorkspace(tree);
    if (!options.project) {
        options.project = Object.keys(workspace.projects)[0];
    }
    const project = workspace.projects[options.project];
    if (!project.architect || !project.architect.build || !project.architect.build.options) {
        return schematics_1.noop();
    }
    const buildOptions = project.architect.build.options;
    if (!buildOptions)
        return schematics_1.noop();
    if (!buildOptions.scripts) {
        buildOptions.scripts = [];
    }
    const scripts = buildOptions.scripts;
    if (!scripts.find(s => s.bundleName === 'polyfill-webcomp-es5')) {
        scripts.push({
            bundleName: 'polyfill-webcomp-es5',
            input: 'node_modules/@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js'
        });
    }
    if (!scripts.find(s => s.bundleName === 'polyfill-webcomp')) {
        scripts.push({
            bundleName: 'polyfill-webcomp',
            input: 'node_modules/@webcomponents/webcomponentsjs/bundles/webcomponents-sd-ce-pf.js'
        });
    }
    workspace_1.updateWorkspace(tree, workspace);
    return schematics_1.noop();
}
function loadPackageJson(tree) {
    const pkg = tree.read('package.json');
    if (pkg === null)
        throw Error('could not read package.json');
    const contentAsString = pkg.toString('UTF-8');
    const config = JSON.parse(contentAsString);
    return config;
}
